### ISSUE TITLE
How to enable automatic opening of parentheses when selecting a function 

### ISSUE THREAD

**Issue by sevaho** [Shougo/deoplete.nvim#608]
**Warning:  I will close the issue without the minimal init.vim and the reproduction instructions.**

# Problems summary

So for example in Go, when typing "fmt." I see the buffer form deoplete with all the functions but when selecting a function for example "Fprintln(w io.Writer, a ...interface{}) (n int, err error)" it completes only to "fmt.Fprintln".

## Expected

What I expect is that it completes to "fmt.Fprintln(|", and with the help of neopairs to "fmt.Fprintln(|)", Where "|" is the cursor.

## Environment Information

 * deoplete version(SHA1): updated

 * OS: Arch Linux, kernel: 4.14.13

 * neovim/Vim version: 0.2.2

 * `:checkhealth` or `:CheckHealth` result(neovim only):

```
health#deoplete#check
========================================================================
## deoplete.nvim
  - OK: has("nvim") was successful
  - OK: has("python3") was successful
  - INFO: If you're still having problems, try the following commands:
    $ export NVIM_PYTHON_LOG_FILE=/tmp/log
    $ export NVIM_PYTHON_LOG_LEVEL=DEBUG
    $ nvim
    $ cat /tmp/log_{PID}
    and then create an issue on github

health#nvim#check
========================================================================
## Configuration
  - OK: no issues found

## Performance
  - OK: Build type: Release

## Remote Plugins
  - OK: Up to date

## terminal
  - INFO: key_backspace (kbs) terminfo entry: key_backspace=^H
  - INFO: key_dc (kdch1) terminfo entry: key_dc=\E[3~
  - INFO: $XTERM_VERSION=''
  - INFO: $VTE_VERSION=''
  - INFO: $TERM_PROGRAM=''
  - INFO: $SSH_TTY=''

health#provider#check
========================================================================
## Clipboard (optional)
  - OK: Clipboard tool found: xsel

## Python 2 provider (optional)
  - INFO: `g:python_host_prog` is not set.  Searching for python2 in the environment.
  - INFO: Multiple python2 executables found.  Set `g:python_host_prog` to avoid surprises.
  - INFO: Executable: /usr/sbin/python2
  - INFO: Other python executable: /usr/bin/python2
  - INFO: Other python executable: /sbin/python2
  - INFO: Other python executable: /bin/python2
  - INFO: Python2 version: 2.7.14
  - INFO: python2-neovim version: 0.2.0
  - OK: Latest python2-neovim is installed: 0.2.0

## Python 3 provider (optional)
  - INFO: `g:python3_host_prog` is not set.  Searching for python3 in the environment.
  - INFO: Multiple python3 executables found.  Set `g:python3_host_prog` to avoid surprises.
  - INFO: Executable: /usr/sbin/python3
  - INFO: Other python executable: /usr/bin/python3
  - INFO: Other python executable: /sbin/python3
  - INFO: Other python executable: /bin/python3
  - INFO: Python3 version: 3.6.4
  - INFO: python3-neovim version: 0.2.0
  - OK: Latest python3-neovim is installed: 0.2.0

## Ruby provider (optional)
  - INFO: Ruby: ruby 2.5.0p0 (2017-12-25 revision 61468) [x86_64-linux]
  - INFO: Host: /home/sevaho/.gem/ruby/2.4.0/bin/neovim-ruby-host

## Provide a minimal init.vim/vimrc with less than 50 lines (Required!)
```

```vim
call plug#begin('~/.local/share/nvim/plugged')

Plug 'Shougo/deoplete.nvim'
Plug 'zchee/deoplete-go', { 'do': 'make'}
Plug 'Shougo/neopairs.vim'

call plug#end()

let g:neopairs#enable = 1
let g:deoplete#enable_at_startup = 1
```

## The reproduce ways from neovim/Vim starting (Required!)

 1. Type fmt.
 2. Deoplete completiong menu shows with functions
 3. select a function, but no parentheses are typed.

## Generate a logfile if appropriate

 1. export NVIM_PYTHON_LOG_FILE=/tmp/log
 2. export NVIM_PYTHON_LOG_LEVEL=DEBUG
 3. nvim -u minimal.vimrc
 4. some works
 5. cat /tmp/log_{PID}

## Screen shot (if possible)
![20180113-133417](https://user-images.githubusercontent.com/10941956/34906193-40f39532-f868-11e7-8635-d16f2d469ce5.png)

**Comment by uinput** [comment]
Did you try the `converter_auto_paren` filter for deoplete?

**Comment by sevaho** [comment]
Hmm I'm sorry but I don't quite understand how you activate it, I have this but I suppose it's wrong:

```
let g:deoplete#converter_auto_paren = 1
```

**Comment by sevaho** [comment]
Ok I got it, but it does not work with neopairs.

```
call deoplete#custom#source('_', 'converters',
		\ ['converter_auto_paren'])
```

**Comment by Shougo** [comment]
Yes.  It is the feature.

You can also use neosnippet.vim.

**Comment by sevaho** [comment]
What? So it is not possible to get autclosed parentheses with deoplete and neopairs?

**Comment by Shougo** [comment]
I have tested the below vimrc.

```vim
set rtp+=~/.cache/dein/repos/github.com/Shougo/deoplete.nvim/
set rtp+=~/.cache/dein/repos/github.com/zchee/deoplete-go/
set rtp+=~/src/neopairs.vim/

let g:deoplete#enable_at_startup = 1
let g:neopairs#enable = 1

call deoplete#custom#source('_', 'converters', ['converter_auto_paren'])
```

It works.

**Comment by Shougo** [comment]
You must select the candidate by `<C-y>`.

**Comment by sevaho** [comment]
Hmm it looks like it is not working in neovim.

my init.nvim:

```
call plug#begin('~/.local/share/nvim/plugged')

Plug 'Shougo/deoplete.nvim'   
Plug 'zchee/deoplete-go', { 'do': 'make'}
Plug 'Shougo/neopairs.nvim'

call plug#end()

let g:deoplete#enable_at_startup = 1
let g:neopairs#enable = 1

call deoplete#custom#source('_', 'converters', ['converter_auto_paren'])
```

No closing of parentheses:

Getting autocomplete targets after writing "fmt.":

![20180115-103131](https://user-images.githubusercontent.com/10941956/34936091-60e5097e-f9e0-11e7-8a71-b48fb0434164.png)

Scrolling trough the list with <C-n>

![20180115-103138](https://user-images.githubusercontent.com/10941956/34935923-d7de22dc-f9df-11e7-99b5-90cd7e774213.png)

Selecting a target with <C-y>:

![20180115-103142](https://user-images.githubusercontent.com/10941956/34935925-db3bbc0a-f9df-11e7-8f79-d1d40fa7dea7.png)

It looks like neopairs isn't doing anything, any idea how to debug?

Also my full nvim folder structure:

```
.
├── [sevaho     40]  autoload
│   ├── [sevaho    72K]  plug.vim
├── [sevaho     28]  colors
│   └── [sevaho    21K]  minimalist.v│   └── [sevaho   2.5K]  templates.vim
im
├── [sevaho    114]  ftdetect
│   ├── [sevaho     58]  dockerfile.vim
│   ├── [sevaho     56]  powershell.vim
│   ├── [sevaho    101]  pug.vim
│   ├── [sevaho     45]  tex.vim
│   └── [sevaho     60]  vagrantfile.vim
├── [sevaho    304]  init.vim
├── [sevaho     46]  plugin
│   ├── [sevaho    22K]  asmx86.vim
├── [sevaho     76]  snippets
│   ├── [sevaho    231]  markdown.snip
│   ├── [sevaho    222]  php.snip
│   ├── [sevaho    611]  python.snip
│   └── [sevaho    153]  _.snip
```

**Comment by Shougo** [comment]
Really?
I cannot reproduce the problem.
You should check neopairs code and CompleteDone handler is called.

**Comment by PendragonSaber** [comment]
Hi @Shougo It is recommended to reopen this question because it is always there。 I also got the same problem.When I typed in, I got `fmt.Println` and didn't get any parentheses.I set`call  deoplete＃custom＃source（' _ '，' converters '，[ ' converter_auto_paren ' ]` But I got the error，Hope to get answers.
like this
![33](https://user-images.githubusercontent.com/41671631/50722559-04d72880-110c-11e9-9ff4-67862de6c402.png)
this is my mininit.vim
```
Plug 'jiangmiao/auto-pairs'
Plug 'Shougo/deoplete.nvim', { 'do': ':UpdateRemotePlugins' }
Plug 'zchee/deoplete-go', { 'do': 'make','for':'go'}
Plug 'Shougo/neosnippet'
Plug 'Shougo/neosnippet-snippets'
" deoplete + neosnippet + autopairs changes
let g:AutoPairsMapCR=0 
let g:deoplete#auto_complete_start_length = 1 
let g:deoplete#enable_at_startup = 1 
let g:deoplete#enable_smart_case = 1 
imap <expr><TAB> pumvisible() ? "\<C-n>" : (neosnippet#expandable_or_jumpable() ? "\<Plug>(neosnippet_expand_or_jump)" : "\<TAB>") 
imap <expr><S-TAB> pumvisible() ? "\<C-p>" : "\<S-TAB>"
imap <expr><CR> pumvisible() ? deoplete#mappings#close_popup() : "\<CR>\<Plug>AutoPairsReturn"
call deoplete#custom#set('_', 'converters', ['converter_auto_paren'])
" deoplete-go settings
let g:deoplete#sources#go#gocode_binary = $GOPATH.'/bin/gocode'
let g:deoplete#sources#go#sort_class = ['package', 'func', 'type', 'var', 'const']
let g:deoplete#sources#go#use_cache = 1
let g:deoplete#sources#go#json_directory = '~/.cache/deoplete/go/'
```

**Comment by PendragonSaber** [comment]
> 你应该重新阅读这个问题......
Sorry, maybe it’s not clear to me. I know this is a feature of `neoparis`. I tried it only when I only had `deoplete` and `deoplete-go`. I found that it is also without parentheses. I didn’t install any parentheses. Plugin,So is there no brackets under the plugin without `neoparis`?
![untitled](https://user-images.githubusercontent.com/41671631/50724968-2696d680-1131-11e9-8f03-1f10df3f799a.gif)
